test/core/slice/slice_test.cc

```diff
@@ -221,73 +227,217 @@ static void test_slice_split_tail_works(size_t length) {
+  EXPECT_EQ(slice.size(), 5);
+  EXPECT_EQ(slice.length(), 5);
+  EXPECT_EQ(slice.as_string_view(), "hello");
+  EXPECT_EQ(0, memcmp(slice.data(), "hello", 5));
+}
+
+class SliceSizedTest : public ::testing::TestWithParam<size_t> {};
+
+std::string RandomString(size_t length) {
+  std::string str;
+  std::random_device r;
+  for (size_t i = 0; i < length; ++i) {
+    str.push_back(char(r()));
+  }
+  return str;
 }
 
-int main(int, char**) {
-  unsigned length;
-  test_slice_malloc_returns_something_sensible();
-  test_slice_new_returns_something_sensible();
-  test_slice_new_with_user_data();
-  test_slice_new_with_len_returns_something_sensible();
-  for (length = 0; length < 128; length++) {
-    test_slice_sub_works(length);
-    test_slice_split_head_works(length);
-    test_slice_split_tail_works(length);
+TEST_P(SliceSizedTest, FromCopiedString) {
+  const std::string str = RandomString(GetParam());
+  Slice slice = Slice::FromCopiedString(str);
+
+  EXPECT_EQ(slice.size(), str.size());
+  EXPECT_EQ(slice.length(), str.size());
+  EXPECT_EQ(slice.as_string_view(), str);
+  EXPECT_EQ(0, memcmp(slice.data(), str.data(), str.size()));
+  for (size_t i = 0; i < str.size(); ++i) {
+    EXPECT_EQ(slice[i], uint8_t(str[i]));
+  }
+
+  EXPECT_TRUE(slice.is_equivalent(slice.Ref()));
+  EXPECT_TRUE(slice.is_equivalent(slice.AsOwned()));
+  EXPECT_TRUE(slice.is_equivalent(slice.Ref().IntoOwned()));
+}
+
+INSTANTIATE_TEST_SUITE_P(SliceSizedTest, SliceSizedTest,
+                         ::testing::ValuesIn([] {
+                           std::vector<size_t> out;
+                           size_t i = 1;
+                           size_t j = 1;
+                           while (i < 1024 * 1024) {
+                             out.push_back(j);
+                             size_t n = i + j;
+                             i = j;
+                             j = n;
+                           }
+                           return out;
+                         }()),
+                         [](const testing::TestParamInfo<size_t>& info) {
+                           return std::to_string(info.param);
+                         });
+
+size_t SumSlice(const Slice& slice) {
+  size_t x = 0;
+  for (size_t i = 0; i < slice.size(); ++i) {
+    x += slice[i];
   }
-  test_slice_from_copied_string_works();
-  test_moved_string_slice();
-  test_string_view_from_slice();
-  return 0;
+  return x;
+}
+
+TEST(SliceTest, ExternalAsOwned) {
+  std::unique_ptr<std::string> external_string(
+      new std::string(RandomString(1024)));
+  Slice slice(ExternallyManagedSlice(external_string->data(),
+                                     external_string->length()));
+  const auto initial_sum = SumSlice(slice);
+  Slice owned = slice.AsOwned();
+  EXPECT_EQ(initial_sum, SumSlice(owned));
+  external_string.reset();
+  // In ASAN (where we can be sure that it'll crash), go ahead and read the
+  // bytes we just deleted.
+#if defined(__has_feature)
```

Comment:
Maybe use `BuiltUnderAsan()` here, since it covers a few more cases?

https://github.com/grpc/grpc/blob/79d684529d9db60aa2c5e82deb43e297a6818cfb/test/core/util/test_config.h#L44

Reply:
Done.